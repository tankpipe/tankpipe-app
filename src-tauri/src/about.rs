use serde::{Serialize, Deserialize};
use data_encoding::BASE64;


const VERSION: &str = env!("CARGO_PKG_VERSION");
const CONTACT: &str = "aHR0cHM6Ly9kaXNjb3JkLmdnL3V6cTg5R1ByVWc=";

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub struct About {
    pub version: String,
    pub contact: String,
}

impl About {
    pub fn new() -> About {
        let input: Vec<u8> = CONTACT.into();
        let binding = BASE64.decode(&input).unwrap();
        let s = String::from_utf8_lossy(&binding);
        About{
            version: VERSION.to_string(),
            contact: s.into_owned()
        }
    }
}
