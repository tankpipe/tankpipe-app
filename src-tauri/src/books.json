{   
    "accounts": {
      "44742ea3-f835-4267-b62e-0c1d4796d575": {
        "id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "name": "Savings Account 1",
        "account_type": "Debit",
        "balance": "0",
        "starting_balance": "0"
      },
      "66ac7995-4991-495a-b44e-d9882d579260": {
        "id": "66ac7995-4991-495a-b44e-d9882d579260",
        "name": "Credit Account 1",
        "account_type": "Credit",
        "balance": "0",
        "starting_balance": "0"
      }
    },
    "scheduled_transactions": [
      {
        "id": "8565ba58-2df8-4690-b4f1-7849b2b4b48f",
        "name": "Some income",
        "period": "Months",
        "frequency": 1,
        "start_date": "2022-06-04",
        "last_date": "2022-06-04",
        "amount": "200",
        "description": "Money in",
        "dr_account_id": "cb946148-79b0-43bc-a3b8-b8bd811aa1ad",
        "cr_account_id": "c77327ef-d27d-434c-a205-2e975e822c87"
      }
    ],
    "transactions": [
      {
        "id": "4ed6200e-4d96-4c40-b140-36de0cfc7bc8",
        "date": "2022-06-04",
        "description": "received moneys",
        "dr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "cr_account_id": "66ac7995-4991-495a-b44e-d9882d579260",
        "amount": "10000",
        "status": "Recorded"
      },
      {
        "id": "f70e1b5d-12ee-47b0-9fbb-ab7671ff4595",
        "date": "2022-06-05",
        "description": "Gave some moneys back",
        "dr_account_id": "66ac7995-4991-495a-b44e-d9882d579260",
        "cr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "amount": "98.99",
        "status": "Recorded"
      },
      {
        "id": "4ed6200e-4d96-4c40-b140-36de0cfc7bc9",
        "date": "2022-07-01",
        "description": "received more moneys",
        "dr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "cr_account_id": "66ac7995-4991-495a-b44e-d9882d579260",
        "amount": "250",
        "status": "Recorded"
      },
      {
        "id": "4ed6200e-4d96-4c40-b140-36de0cfc7bd1",
        "date": "2022-07-01",
        "description": "received more moneys",
        "dr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "cr_account_id": null,
        "amount": "150.45",
        "status": "Recorded"
      }
    ]
  }