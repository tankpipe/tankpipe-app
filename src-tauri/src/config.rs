use std::{ffi::{OsString, OsStr}, path::PathBuf};
use serde::{Deserializer, Serializer, Serialize, Deserialize};

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub struct FileDetails {
    pub name: String,
    #[serde(serialize_with = "serialize_osstring")]
    #[serde(deserialize_with = "deserialize_osstring")]
    pub path: OsString,
}

impl FileDetails {
    pub fn empty() -> FileDetails {
        FileDetails{ name: "".to_string(), path: OsString::from("") }
    }

    pub fn from_path(name: &str, path: PathBuf) -> FileDetails {
        FileDetails{ name: name.to_string(), path: path.into_os_string() }
    }

    pub fn from_os_string(name: &str, path: OsString) -> FileDetails {
        FileDetails{ name: name.to_string(), path: path }
    }
}

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub enum DateFormat {
    Locale,
    Regular,
    US,
    ISO,
}

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub struct Config {
    pub version: String,
    #[serde(serialize_with = "serialize_osstring")]
    #[serde(deserialize_with = "deserialize_osstring")]
    pub data_dir: OsString,
    #[serde(serialize_with = "serialize_osstring")]
    #[serde(deserialize_with = "deserialize_osstring")]
    pub config_dir: OsString,
    pub last_file: FileDetails,
    pub recent_files: Vec<FileDetails>,
    pub display_date_format: DateFormat,
    pub import_date_format: String
}

impl Config {
    pub fn data_dir_as_os_str(&self) -> &OsStr {
        &self.data_dir.as_os_str()
    }

    pub fn settings_path(&self) -> PathBuf {
        PathBuf::from(self.config_dir.clone()).join("settings.json")
    }

    pub fn file_path(&self, name: &str) -> PathBuf {
        PathBuf::from(self.data_dir.clone()).join(name)
    }

    pub fn set_last_from_path(&mut self, path: OsString, name: &str) {
        let last_file: FileDetails;
        let os_string_path = OsString::from(path);
        self.last_file.path = os_string_path.clone();
        self.last_file.name = name.to_string();
        let index = self.recent_files.iter().position(|f| *f.path == os_string_path);
        match index {
            Some(i) => {
                last_file = self.recent_files.get(i).unwrap().clone();
                self.recent_files.remove(index.unwrap());
            },
            None =>last_file = FileDetails::from_os_string(name, os_string_path)
        }
        self.recent_files.insert(0, last_file);
    }

    pub fn set_last(&mut self, last_file: FileDetails) {
        if let Some(index) = self.recent_files.iter().position(|f| *f.path == last_file.path) {
            self.recent_files.remove(index);
        }
        self.last_file = last_file.clone();
        self.recent_files.insert(0, last_file);
    }

}

pub fn deserialize_osstring<'de, D>(deserializer: D) -> Result<OsString, D::Error>
    where D: Deserializer<'de>
{
    let path = String::deserialize(deserializer)?;
    Ok(OsString::from(path))
}

pub fn serialize_osstring<S>(path: &OsString, serializer: S) -> Result<S::Ok, S::Error>
    where S: Serializer,
{
    serializer.serialize_some(&PathBuf::from(path.as_os_str()))
}