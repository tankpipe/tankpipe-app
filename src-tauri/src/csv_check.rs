use serde::{Serialize, Deserialize};


use crate::reader::ColumnTypes;

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct CsvCheck {
    pub column_types: ColumnTypes,
    pub header_row: Vec<String>,
    pub sample_rows: Vec<Vec<String>>,
    pub mapping_exists: bool,
}

impl CsvCheck {
    pub fn create_new(column_types: ColumnTypes, header_row: Vec<String>, sample_rows: Vec<Vec<String>>) -> CsvCheck {
        CsvCheck {
            column_types: column_types,
            header_row: header_row,
            sample_rows: sample_rows,
            mapping_exists: false,
        }
    }

    pub fn set_mapping_exists(self, mapping_exists: bool) -> CsvCheck {
        CsvCheck {
            column_types: self.column_types,
            header_row: self.header_row,
            sample_rows: self.sample_rows,
            mapping_exists: mapping_exists,
        }
    }
}


