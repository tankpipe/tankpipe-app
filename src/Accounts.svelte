<script>
    import EditAccount from "./EditAccount.svelte"
	import Icon from '@iconify/svelte';
    import { modes, views } from "./context";

    export let curAccount
    export let accounts
    export let loadAccounts
	export let context

	let ACCOUNT_TYPES = {
		Asset: "Assets",
		Liability: "Liabilities",
		Revenue: "Revenues",
		Expense: "Expenses",
		Equity: "Equity"
	}
    let mode = "ACCOUNTS"
    let editMode = "ADD"
	let lastAccountType = "";

	$: {
		if (accounts)
			if (accounts.length < 1) {
				mode = "EDIT"
			}
	}

	const close = () => {
        console.log("close")
		context.setView(views.ACCOUNTS)
        mode = "ACCOUNTS";
    }

    const handleAddClick = () => {
		curAccount = null
        editMode = "ADD"
		mode = "EDIT"
	}

    const selectAccount = (account) => {
		console.log(account)
        curAccount = account
		console.log(curAccount.id)
		context.setView(views.TRANSACTIONS, modes.LIST)
    }

	const editAccount = (account) => {
        curAccount = account
        editMode = "EDIT"
		mode = "EDIT"
        console.log("selected: " + curAccount.name);
    }

	const checkAccountType = (account) => {
		if (account.account_type !== lastAccountType) {
			lastAccountType = account.account_type
			return true
		}
		return false
	}

</script>

<div class="account-heading">
	{#if mode === "ACCOUNTS"}
	<div class="toolbar"><div class="toolbar-icon" on:click="{handleAddClick}" title="Create a new account"><Icon icon="mdi:plus-box-outline"  width="24"/></div></div>
	{/if}
</div>


{#if mode === "EDIT"}
<EditAccount {curAccount} {loadAccounts} {close} {editMode} initialize={accounts.length < 1}/>
{/if}
{#if mode === "ACCOUNTS"}
<div class="form-heading">Accounts</div>
<div class="scroller">
	<div class="accounts">
    {#each accounts as a}
	{#if checkAccountType(a)}
		<div class="account-type">{ACCOUNT_TYPES[a.account_type]}</div>
	{/if}
        <div class="card" on:click={() => selectAccount(a) }>{a.name}<div class="edit-icon" on:click={() => editAccount(a) }><Icon icon="mdi:pencil" /></div></div>
    {/each}
	</div>
</div>
{/if}
<style>

	.scroller{
		height: 100%;
		width: 100%;
		overflow: scroll;
		margin-top: 15px;
	}

	.accounts {
		margin-right: 20px;
	}

	.account-type {
		font-size: 0.8em;
        font-weight: 500;
        color: #757575;
		margin: 5px 0px -5px 10px;
		float: left;
		clear: both;
	}

	.edit-icon {
		float: right;
		color: #524e4e;
	}

	.card:hover .edit-icon {
		color: #666;
	}

	.edit-icon:hover {
		color: #C0C0C0 !important;
	}

    .card {
		float: left;
		clear: both;
		margin: 10px;
		background-color: #524e4e;
		padding: 10px;
		border-radius: 10px;
		color: #E0E0E0;
		min-width: 300px;
		text-align: left;
	}

	.card:hover {
		cursor: pointer;
		color: #FFF;
	}

</style>