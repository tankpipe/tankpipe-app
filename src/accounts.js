import { writable } from 'svelte/store'

const accounts = writable({})

export {accounts as accounts}