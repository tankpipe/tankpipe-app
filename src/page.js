import { writable } from 'svelte/store'

const views = {
    BOOKS: "BOOKS",
    ACCOUNTS: "ACCOUNTS",
    TRANSACTIONS: "TRANSACTIONS",
    SCHEDULES: "SCHEDULES",
    SETTINGS: "SETTINGS",
    NET_ASSETS: "NET_ASSETS",
    JOURNAL: "JOURNAL",
}

const modes = {
    NEW: "NEW",
    EDIT: "EDIT",
    MULTI_EDIT: "MULTI_EDIT",
    LIST: "LIST",
    LOAD: "LOAD"
}

const page = writable({view: views.ACCOUNTS, mode:modes.LIST, payload:{}})

const isEditMode = (page) => {
    return page.mode === modes.EDIT || page.mode === modes.NEW || page.mode === modes.MULTI_EDIT
}

const isSingleEditMode = (page) => {
    return page.mode === modes.EDIT || page.mode === modes.NEW
}

const isMultiEditMode = (page) => {
    return page.mode === modes.MULTI_EDIT
}

const isListMode = (page) => {
    return page.mode === modes.LIST
}

export {page, views, modes, isEditMode, isMultiEditMode, isSingleEditMode, isListMode}