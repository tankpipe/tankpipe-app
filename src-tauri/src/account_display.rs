use accounts::account::{AccountType, Account};
use chrono::NaiveDate;
use rust_decimal::Decimal;
use serde::{Serialize, Deserialize};
use uuid::Uuid;
use accounts::serializer::*;
use rust_decimal_macros::dec;

#[derive(Clone, Serialize, Deserialize)]
pub struct NewAccount {
	pub name: String,
	pub account_type: AccountType,
	pub starting_balance: Decimal
}

impl NewAccount {
	pub fn to_account(self) -> Account {
		Account {
			id: Uuid::new_v4(),
			name: self.name,
			starting_balance: self.starting_balance,
			account_type: self.account_type,
            balance: dec!(0)
		}
	}

}


#[derive(Clone, Serialize, Deserialize)]
pub struct DateParam {
    #[serde(serialize_with = "serialize_naivedate")]
    #[serde(deserialize_with = "deserialize_naivedate")]
	pub date: NaiveDate
}

