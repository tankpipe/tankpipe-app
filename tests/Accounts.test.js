import { render, waitFor } from '@testing-library/svelte'
import Accounts from '../src/Accounts.svelte'
import {accounts} from '../src/accounts.js'
import {page, views, modes} from '../src/page'
import account_data from './account_data.json'

accounts.set(account_data)

it('is displayed correctly', async () => {
    page.set({view: views.ACCOUNTS, mode: modes.LIST})
    const mockFetchAccounts = jest.fn(() => Promise.resolve(account_data));
    global.invoke = mockFetchAccounts;

    const {container} = render(Accounts, {loadAccounts: mockFetchAccounts, curAccount: account_data[0]})
    //const _waitForRenderUpdate = await findByTitle('Assets')
    expect(container.outerHTML).toMatchSnapshot();
});

