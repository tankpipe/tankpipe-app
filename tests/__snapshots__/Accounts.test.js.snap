// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`is displayed correctly 1`] = `"<div><!----><!----><div class="message">To get started give your first set of accounts a name. For example, Personal Finances.<p></p></div><!----> <div class="form svelte-1ozwy1d"><div class="form-heading">New Books</div> <div class="form-row svelte-1ozwy1d"><div class="widget svelte-1ozwy1d"><label for="name">Name</label> <input id="name" class="description-input svelte-1ozwy1d"></div></div> <div class="form-button-row svelte-1ozwy1d"><div class="msg-panel svelte-1ozwy1d"><!----> <!----></div> <div class="widget buttons svelte-1ozwy1d"><!----> <button class="svelte-1ozwy1d">Add</button></div></div></div></div>"`;
